include/rosparam_shortcuts/parameter.h

```diff
@@ -0,0 +1,245 @@
+NodeParameter::declare() override{ /* returns shared_from_this */ };
+
+/*
+ * Maintains all Parameter instances and does the heavy lifting with actually declaring and interfacing with dynamic
+ * rclcpp Parameters. Provides additional convenience functions for validating all parameters, printing, etc...
+ */
+class NodeConfig
+{
+public:
+	NodeConfig(const rclcpp::Node::SharedPtr& node, const std::string base_name = "");
+
+protected:
+	/*
+	 * Declare Parameter types, internally calls Parameter<ParameterT>::declare() to get
+	 * all actual node parameters which are then declared using the node parameter interface.
+	 */
+	template <typename ParameterT>
+	Parameter<ParameterT>& declare(Parameter<ParameterT>& parameter);
+
+	/*
+	 * Undeclare a parameter, remove all callbacks that have been registered using the declare() call
+	 */
+	template <typename ParameterT>
+	bool undeclare(const Parameter<ParameterT>& parameter);
+	bool undeclare(const std::string& name);
+
+	/*
+	 * Return the parameter for a given name, if not found parameter is nullptr
+	 */
+	// Not possible when working with references, maybe only for actual node parameters?
+	// std::shared_ptr<Parameter<ParameterT>> get(const std::string& name);
+	// std::shared_ptr<const Parameter<ParameterT>> get(const std::string& name) const noexcept;
+
+	//
+	// TODO: insert utility functions:  validate(), print, ostream<<, errors(), warnings()...
+	//
+
+private:
+	const rclcpp::Node::SharedPtr node_;
+	const std::string base_name_;
+
+	// Class similar to the one in rosparam_shortcuts/node_parameters.h
+	// implements ROS calls/checks/callbacks for node_parameters_
+	NodeParametersInterface node_parameters_interface_;
+
+	// Store parameters, only node parameters (rclcpp::Parameters) are stored inside NodeConfig
+	std::map<std::string, Parameter&> parameters_;
+	std::map<std::string, std::shared_ptr<NodeParameter>> node_parameters_;
+
+	// Needed? keep track of registered callbacks that are required for linking SetParameters requests or
+	// exposing high-level update changes (invalid/valid states)
+	// std::string<std::string, std::string> registered_node_parameter_callbacks_ids_;
+	// std::string<std::string, std::string> registered_parameter_callbacks_ids_;
+}
+
+struct ExampleConfig : public NodeConfig
+{
+	ExampleConfig(const rclcpp::Node::SharedPtr& node) : NodeConfig(node) {
+		// declare simple parameters just like that
+		declare(int1_param);
+		declare(int2_param);
+		declare(int3_param);
+
+		// "double_param", add some description and validity checks
+		declare(double_param)
+		    .describe("This is a very important double parameter", "should be greater than 0 and not 7")
+		    .warnIf([](const double& val, std::string& message) {
+			    message = "Value is 7, are you sure this is a good value?";
+			    return val == 7;
+		    })
+		    .errorIf([](const double& val, std::string& message) {
+			    message = "Value must not be negative!";
+			    return val < 0;
+		    });
```

Comment:
I really like this builder syntax, the only thing I think it is missing is a default value.  Maybe there would be a `default(ParameterT )` method?  There should also be support for allowing parameters to be unset (no default) but waiting for some timeout for them to be set.  This makes sense because the way nodes are launched the parameters are sent to the node after they are started so they are always unset and it is possible to declare a parameter before it is set and then the parameter service interface will set it.

Reply:
The default value can be set with the Parameter constructor: `Parameter<double> double_param {"double param", 2.0 };`.
For non-trivial types this becomes a little more difficult. Having the parameter name and default value part of that builder syntax could improve readability. How about initializing all parameters with default constructors and just setting the name and possible default with `declare()`. Undeclared parameters would simply throw an exception if they are used.

```suggestion
		declare("double_param", double_param, 2.0 /* default */)
		    .describe("This is a very important double parameter", "should be greater than 0 and not 7")
		    .warnIf([](const double& val, std::string& message) {
			    message = "Value is 7, are you sure this is a good value?";
			    return val == 7;
		    })
		    .errorIf([](const double& val, std::string& message) {
			    message = "Value must not be negative!";
			    return val < 0;
		    });
```